### ISSUE TITLE
Shorthand CORS notation is useless for GET, HEAD and POST methods.

### ISSUE THREAD

**Issue by shpoont** [aws/serverless-application-model#387]
Example code:
```
Globals:
  Api:
    Name: test
    Cors: "'*'"
Resources:
  TestFunction:
    Type: AWS::Serverless::Function
    Properties:
      ...
      Events:
        PostResource:
          Type: Api
          Properties:
            Path: /test
            Method: post
```

The resulting API configuration does not support [preflight requests](https://www.w3.org/TR/cors/#resource-preflight-requests) for GET, HEAD or POST methods.

In addition, since CORS configuration for [simple requests](https://www.w3.org/TR/cors/#resource-requests) can not be done in Lambda proxy integration, but has to be implemented in actual Lambda code, specifying  `Cors: "'*'"` is useless for HTTP methods mentioned above.

**Comment by Jun711** [comment]
@rabowskyb 
```
 AllowHeaders:
        - "'customHeader1'"
        - "'customHeader2'"
```
To allow a specific list of headers, is the above syntax correct? Thanks

**Comment by rabowskyb** [comment]
That looks like it should work, I'd suggest trying it out.  If it doesn't work when tested, I'd recommend opening a separate issue in this repo.

**Comment by brettstack** [comment]
@shpoont could you clarify what you mean by "useless"? What was the expected and actual results?

**Comment by shpoont** [comment]
@brettstack sure, by "useless" I mean that when I use shorthand CORS notation, the resources created do not match the specs defined by w3.org, and requests to those resources are being blocked by browsers.

**Comment by bikerp** [comment]
I'm facing the same issue. In order to make CORS work I have to return Access-Control-Allow-Origin header directly from Lambda function.

**Comment by brettstack** [comment]
That's right. SAM can only create and configure an OPTIONS method for you to perform the CORS preflight. It can't modify the response from the Lambda Function as API Gateway does not allow for response mappings from Proxy Integrations.

**Comment by abbottdev** [comment]
@brettstack  - I *think* the hinted at feature here is that if people specify CORS configuration in their cfn templates, the expectation would be that those CORS headers are appended/merged to the output?

I feel that would be a cleaner way of developing - imagine you have nested stacks and the URL of a deployed client app is in an S3 bucket - that URL is output from another stack - The cloudformation template "feels" like that's where the CORS domain parts live as that's where you set things like the base URL.

It is possible to workaround, by exposing an environment variable in the Lambda - but this feels like you're exposing something for the sake of cloudformation. It would be nice for it to work either way IMO 😄

**Comment by brettstack** [comment]
That's not possible. API Gateway limitations.

**Comment by abbottdev** [comment]
Ah man - thanks for the info regardless @brettstack

**Comment by c2tarun** [comment]
Based on the discussion above we cannot add CORS header to the response due API Gateway not supporting response mapping for Proxy Integrations. I am closing this issue for now. Please feel free to re-open if you have more questions.
